src/Common/AbstractEnum.php

```diff
@@ -0,0 +1,356 @@
+                sprintf('%s is not a valid backing value for enum %s', (string) $value, static::class)
+            );
+        }
+        return $instance;
+    }
+
+    /**
+     * Try to create an enum instance from a value, returns null if invalid
+     *
+     * @param string|int $value The enum value
+     * @return static|null
+     */
+    public static function tryFrom($value): ?self
+    {
+        $constants = self::getConstants();
+        foreach ($constants as $name => $constantValue) {
+            if ($constantValue === $value) {
+                return self::getInstance($constantValue, $name);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get all enum cases
+     *
+     * @return static[]
+     */
+    public static function cases(): array
+    {
+        $cases = [];
+        $constants = self::getConstants();
+        foreach ($constants as $name => $value) {
+            $cases[] = self::getInstance($value, $name);
+        }
+        return $cases;
+    }
+
+    /**
+     * Check if this enum has the same value as the given value
+     *
+     * @param string|int|self $other The value or enum to compare
+     * @return bool
+     */
+    public function equals($other): bool
+    {
+        if ($other instanceof self) {
+            return $this->is($other);
+        }
+
+        return $this->value === $other;
+    }
+
+    /**
+     * Check if this enum is the same instance type and value as another enum
+     *
+     * @param self $other The other enum to compare
+     * @return bool
+     */
+    public function is(self $other): bool
+    {
+        return $this === $other; // Since we're using singletons, we can use identity comparison
+    }
+
+    /**
+     * Get all valid values for this enum
+     *
+     * @return array<string, string|int>
+     */
+    public static function getValues(): array
+    {
+        return self::getConstants();
+    }
+
+    /**
+     * Check if a value is valid for this enum
+     *
+     * @param string|int $value The value to check
+     * @return bool
+     */
+    public static function isValidValue($value): bool
+    {
+        return in_array($value, self::getValues(), true);
+    }
+
+    /**
+     * Create an enum instance from a value (deprecated, use from() instead)
+     *
+     * @param string|int $value The enum value
+     * @return static
+     * @throws InvalidArgumentException If the value is not valid
+     * @deprecated Use from() method instead
+     */
+    public static function fromValue($value): self
+    {
+        return self::from($value);
+    }
+
+    /**
+     * Get or create a singleton instance for the given value and name
+     *
+     * @param string|int $value The enum value
+     * @param string $name The constant name
+     * @return static
+     * @phpstan-return static
+     */
+    private static function getInstance($value, string $name): self
+    {
+        $className = static::class;
+        $key = $name;
```

Comment:
Nit-pick: Seems unnecessary to have a `$key` variable here? Could just use `$name` throughout.

Reply:
Hah! Amazing. Resolved in ebe9d2a